### PR TITLE
fix(scales): bisect correctly on continuous scales

### LINKED ISSUE #221: Handle discontinuous bar series crosshair interactions
In #200, we use d3's `bisectLeft` to calculate the value returned by `invertWithStep`; this helps solve the scenarios where we have discontinuous data in a line or area series, but also now means that for a bar series with discontinuous data, it will snap the band highlighter to the nearest x value with a datum defined.  This can look quite odd for a bar series with a huge gap between x values as below:

bar_chart_hover

(With the previous implementation, hovering over an area with no data in a bar serious would highlight the empty band with no selected data.)

**Is your feature request related to a problem? Please describe.**
It would be nice to either bring back the previous way of handling these empty zones or handle the hover state such that if the cursor is not within a minimum range near a data point, either we show the band highlighter with no data or there is no hover interaction at all.

**Describe the solution you'd like**
Would like to discuss the options for how to handle these cases before beginning on implementatiion 😄 

**Describe alternatives you've considered**

**Additional context**

**Kibana Cross Issues**

**Checklist**

- [x] this request is checked against already exist requests
- [ ] every related Kibana issue is listed under `Kibana Cross Issues` list
- [ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### LINKED ISSUE #227: Crosshairs hover tooltip & highlighter band mispositioned when scale is Linear
**Describe the bug**
When hovering over a clustered series, if the scale type is Scale.Linear, then the hover tooltip and highlighter appear for the next cluster when you hover past the center part of the band.

bars_hover

**To Reproduce**
1. Create a chart with clustered bars (either multiple BarSeriesSpecs or single BarSeriesSpec with splitSeriesAccessors) with `xScaleType={ScaleType.Linear}`.
2. Hover over first band, notice that as you hover over past the halfway point but still in the first band, the tooltip & band highlighter appear for the second band, not the first.

Note that if you change the scale type to `ScaleType.Ordinal`, the hover state appear correctly positioned.

**Expected behavior**
Tooltip & band positioning should match how they appear correctly for Scale.Ordinal.

bars_hover_correct

**Screenshots**
See above

**Version (please complete the following information):**
 - OS: all
 - Browser: all
 - Elastic Charts: v4.2.6

**Additional context**

**Errors in browser console**
None

**Kibana Cross Issues**
None

**Checklist**
- [x] every related Kibana issue is listed under `Kibana Cross Issues` list
- [x] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### LINKED ISSUE #347: Missing annotations on missing ticks
**Describe the bug**
The current implementation is linked to the number of ticks displayed on the chart. If the chart displays a lot of elements, it's possible that annotations placed on the missing ticks are hidden.

**To Reproduce**
1. Use the following chart on the playground:
```
<Chart>
        <Settings
          tooltip={{ type: 'vertical' }}
          debug={false}
          legendPosition={Position.Right}
          showLegend={true}
          rotation={0}
        />
        <Axis
          id={getAxisId('timestamp')}
          title="timestamp"
          position={Position.Bottom}
          tickFormat={niceTimeFormatter([1555819200000, 1555905600000])}
        />
        <Axis id={getAxisId('count')} title="count" position={Position.Left} tickFormat={(d) => d.toFixed(2)} />
        <LineAnnotation
          annotationId={getAnnotationId('annotation1')}
          domainType={AnnotationDomainTypes.XDomain}
          dataValues={[
            {
              dataValue: KIBANA_METRICS.metrics.kibana_os_load[1].data[1][0],
              details: 'tooltip 1',
            },
            {
              dataValue: KIBANA_METRICS.metrics.kibana_os_load[1].data[9][0],
              details: 'tooltip 2',
            },
          ]}
          hideLinesTooltips={true}
          marker={<Icon type="alert" />}
        />
        <BarSeries
          id={getSpecId('dataset A')}
          xScaleType={ScaleType.Time}
          yScaleType={ScaleType.Linear}
          data={KIBANA_METRICS.metrics.kibana_os_load[1].data.slice(0, 50)}
          xAccessor={0}
          yAccessors={[1]}
          stackAccessors={[0]}
        />
      </Chart>
```
2. You will notice a chart without annotations
3. Change the annotation `dataValue` to `...kibana_os_load[1].data[0][0]` and `...kibana_os_load[1].data[10][0]`
4. See the annotation present

**Expected behavior**
The line annotations and rect should be present independently from the ticks.

**Screenshots**
**With annotation on place with no ticks**
Screenshot 2019-08-25 at 23 37 38

**with annotation on place with ticks**
Screenshot 2019-08-25 at 23 37 49

**Version (please complete the following information):**
 - OS: all
 - Browser: all
 - Elastic Charts: master

**Additional context**
n/a

**Errors in browser console**
n/a

**Kibana Cross Issues**
n/a

**Checklist**
- ~[ ] every related Kibana issue is listed under `Kibana Cross Issues` list~
- ~[ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present~

### PR DESCRIPTION
**Pull request by markov00**
## Summary

fix #227, fix #221, fix #347

This commit fix the wrong behaviour of the current invertWithStep implementation.

Aug-25-2019 22-25-39

Aug-25-2019 22-26-52

It also fix the behaviour of cursor sync between charts with only lines and charts with only bars and empty intervals.

Aug-25-2019 22-25-02

It finally fix also the missing line/rect annotation that are dependent on ticks to be shown. The PR removed the dependency with the ticks and allows us to place correctly the annotations depending on chart type

**Notes**
I've the impression that we need an absolute refactoring of the current scale system for linear band scales. The current implementation is a bit hacky because it's a mix of a linear scale adapted to work as a band/ordinal scale.
Right now these two scales differs by 1 pixel, it's not noticeable but it create a bit of confusion on the code and testing.

The playground is updated with an example of multi-chart synched with various configurations

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- ~[ ] Proper documentation or storybook story was added for features that require explanation or tutorials~
- [x] Unit tests were updated or added to match the most common scenarios
- [x] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/chart_types/xy_chart/store/chart_state.ts [hunk 1]
```diff
@@ -664,8 +686,8 @@ export class ChartStore {
       return;
     }
     this.isBrushing.set(false);
-    const minValue = start.x < end.x ? start.x : end.x;
-    const maxValue = start.x > end.x ? start.x : end.x;
+    const minValue = Math.min(start.x, end.x);
```
Comment (nickofthyme) [inline src/chart_types/xy_chart/store/chart_state.ts]: 👍

File: src/utils/themes/light_theme.ts [hunk 2]
```diff
@@ -101,7 +101,7 @@ export const LIGHT_THEME: Theme = {
   },
   crosshair: {
     band: {
-      fill: 'lightgray',
+      fill: 'gray',
```
Comment (nickofthyme) [inline src/utils/themes/light_theme.ts]: Is this something Caroline wants?

**Comment by markov00** [comment]
> Tested locally, looks good!
> 
> I did notice some strange behaviour with one of the charts and the legend toggle 👇
> Screen Recording 2019-08-26 at 08 09 AM
> 
> Have you seen this before?

Yes it's visible on https://elastic.github.io/elastic-charts/?path=/story/mixed-charts--bar-and-lines if you hide the bar

File: src/utils/themes/light_theme.ts (hunk 2 above)
Comment (markov00) [reply src/utils/themes/light_theme.ts]: let me go back to lightgray :)

**Pull request merged**